ducktape/services/service_registry.py

```diff
@@ -0,0 +1,63 @@
+# Copyright 2014 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ducktape.services.service import Service
+from ducktape.utils.local_filesystem_utils import mkdir_p
+
+from collections import OrderedDict
+
+import os
+
+
+class ServiceRegistry(OrderedDict):
+
+    def allocate_nodes(self):
+        """Make all registered services request nodes from the cluster."""
+        for service in self.values():
+            service.allocate_nodes()
+
+    def start_all(self):
+        """Start all currently registered services in the same order in which they were added."""
+        for service in self.values():
+            service.start()
+
+    def stop_all(self):
+        """Stop all currently registered services in the reverse of the order in which they were added.
+
+        Note that this does not clean up persistent state or free the nodes back to the cluster.
+        """
+        for service in reversed(self.values()):
+            try:
+                service.stop()
+            except Exception as e:
```

Comment:
You might want the exceptions you're catching here to be BaseException. There are a few that won't be caught by Exception. See https://docs.python.org/2/library/exceptions.html#exception-hierarchy

Actually, KeyboardInterrupt in particular might need careful thought and handling -- it's unlikely we want to kill the process immediately in most cases, but we'll probably need to coordinate a couple of things. When shutting down a test, we'd probably like for that shutdown to finish (so we need to handle the exception). But then the test runner ideally wouldn't finish running any tests that were left. Might just want to change these to BaseException now and file a ticket for graceful ctrl-c handling.